crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -324,6 +252,43 @@ impl NotifyStream {
             }
         }
     }
+
+    fn get_full_path_and_mask(
+        event: &Event<OsString>,
+        files: Vec<EventDescription>,
+    ) -> Result<(String, FileEvent), NotifyStreamError> {
+        // Unwrap is safe here because event will always contain a file name.
+        let fname = event.name.as_ref().unwrap();
+        // Check if file under watch. If so, then return the full path to the file and the event mask.
+        for file in &files {
+            if let Some(wfname) = file.file_name.as_ref() {
+                if wfname.eq(&fname.to_string_lossy()) {
+                    for mask in &file.masks {
+                        if mask.eq(&event.mask.try_into()?) {
```

Comment:
Here the `cross verification` of the raised event is done to check if the event raised contains the `file name` and the `event masks` that were supposed to be watched for.

Reply:
Using a `Vec<EventDescription>` is not the appropriate struct then.

What you need is a `HashMap<<Option<String>, EventDescription>` since you are always looking for *the* event that matches either some file name (the name of the file where something happened) or none (meaning any file in that directory).

Reply:
I think the proposed data struct does not work because 

- When the same file name can be part of two directories.
- Also when it's `None`, then which one to pick up?

Reply:
I think you missed my point.

I propose to use a `HashMap<Option<String>, EventDescription>` for *each* directory under watch.

```
description: HashMap<c_int, HashMap<Option<String>, EventDescription>>,
```

Look the current code for `get_full_path_and_mask` that use a `Vec<EventDescription>`.
* This code iterates over the event descriptions to find *the* description, if any, that match the current event file. Using a map from file names to event descriptions, this can be done directly. 
* The current code also iterates over the event descriptions to find *the* event description attached to the directory, if any, i.e the entry with *no* filename. This is for that use case that the keys of the map must be options.